Integrate fine-tuned BERT-based MITRE TTP classifier from Hugging Face

zandanyd opened the pull request:
Fine-tuned a BERT model for multi-label classification of MITRE ATT&CK TTPs using TRAM dataset.

Exported the model as a TorchScript (model_scripted.pt) and uploaded it to Hugging Face Hub.

Updated TramClassifierExtractor to download model and tokenizer directly from the Hugging Face repo.

Extended classification logic to process blog content, extracted keywords, and QnA answers as input chunks.

Added a tram_classifier_node to the pipeline that performs MITRE TTP classification using the Hugging Face model.

Added support for .env variable SKIP_MITRE_CLASSIFICATION to optionally disable classification.

zandanyd committed:
Add TRAM classification model files

- Added TorchScript model (`model_scripted.pt`) for MITRE ATT&CK multi-label classification
- Included tokenizer files (`vocab.txt`, `tokenizer_config.json`, `special_tokens_map.json`)
- Added model configuration (`config.json`)
- Added label binarizer (`label_binarizer.pkl`) for decoding model output
- Included MITRE ATT&CK mapping (`enterprise-attack.json`)

These files are required to support automated MITRE TTP classification in the pipeline.

zandanyd committed:
Integrate fine-tuned BERT-based MITRE TTP classifier from Hugging Face

Fine-tuned a BERT model for multi-label classification of MITRE ATT&CK TTPs using TRAM dataset.

Exported the model as a TorchScript (model_scripted.pt) and uploaded it to Hugging Face Hub.

Updated TramClassifierExtractor to download model and tokenizer directly from the Hugging Face repo.

Extended classification logic to process blog content, extracted keywords, and QnA answers as input chunks.

Added a tram_classifier_node to the pipeline that performs MITRE TTP classification using the Hugging Face model.

Added support for .env variable SKIP_MITRE_CLASSIFICATION to optionally disable classification.

zandanyd committed:
#Sort the classifications by confidence of the model and handle errors

zandanyd committed:
handel The large text from the blogs and split it to chunks with Microsoft bling fire

zandanyd committed:
Merge branch 'barvhaim:master' into mitre_ttp

zandanyd committed:
Evaluation

zandanyd committed:
remove evaluation's printings

barvhaim commented:
@zandanyd do we have `transformers` in the requirements.txt? make sure the libs you are using are installed in the package.. have you tried to run the docker container?

barvhaim commented on backend/extractors/tram_classifier_extractor.py:
```diff
@@ -0,0 +1,127 @@
+        return self.enrich_ids_with_metadata(sorted_labels)
+
+    def _split_text_into_chunks(self, text, max_tokens=512, stride=256):
+        sentences = blingfire.text_to_sentences(text).split('\n')
```
we already use `LangChain` so consider using https://pypi.org/project/langchain-text-splitters

barvhaim replied:
I mean from langchain.text_splitter

barvhaim commented on backend/extractors/tram_classifier_extractor.py:
```diff
@@ -0,0 +1,127 @@
+        self.model_path = model_repo
+        self.tokenizer = BertTokenizer.from_pretrained(self.model_path)
+
+        scripted_model_path = hf_hub_download(repo_id=model_repo, filename="model_scripted.pt")
```
Make sure the class is not initialized if the ENV flag is OFF so we don't download it

barvhaim commented on backend/extractors/tram_classifier_extractor.py:
```diff
@@ -0,0 +1,127 @@
+        chunks = []
+
+        # 1. Split article
+        if len(self.tokenizer.encode(self.article_content, add_special_tokens=False)) > 512:
```
if the model config is available, you can get the 512 from it instead of hardcoding.. (usually there is a model config json file)

barvhaim commented on cli.py:
```diff
@@ -90,6 +90,25 @@ def _display_results(console: Console, res: dict, url: str):
     else:
         console.print("No IoCs found\n")
 
+    # TRAM MITRE Classification
```
Enable this part if the ENV var is set

barvhaim commented on backend/pipeline/state.py:
```diff
@@ -10,4 +10,6 @@ class PipelineState(TypedDict):
     qna: List[Dict]
     keywords_found: List[str]
     iocs_found: List[IOC]
+    mitre_attacks: List[Dict]
```
rename to mitre_ttps

barvhaim commented:
@zandanyd looks good, overall comments / checks:
- make sure all required libs are in the requirements.txt
- rename to `mitre_ttps`
- check if you can use text splitters that built in LangChain instead of brining another lib
- check if you can use the transformers to load and use the model instead of PyTorch code, they should have a simpler code with pipelines https://huggingface.co/docs/transformers/en/pipeline_tutorial
- make sure to not loading the model if the ENV var is not set

zandanyd committed:
- Renamed class to `mitre_ttps` for consistency and clarity
- Replaced `blingfire` sentence splitter with `RecursiveCharacterTextSplitter` from LangChain
  - Token counting adapted using HuggingFace tokenizer
- Explicitly cache HuggingFace model files to ~/.cache/mitre_model
  - Prevents redundant downloads and improves Docker performance
- Load model and related files only if ENV var is enabled (e.g., USE_MITRE_TTP=true)
  - Skips downloading and loading heavy files if not needed
- Added missing dependencies to `requirements.txt`:
  - `transformers`, `huggingface_hub`, `joblib`, `langchain`, `torch`
- Docker run verified
